google-cloud-spanner/src/main/java/com/google/cloud/spanner/Interval.java

```diff
@@ -0,0 +1,316 @@
+    BigDecimal seconds = new BigDecimal(getNullOrDefault(matcher, 7).replace("S", ""));
+
+    long totalMonths = years * MONTHS_PER_YEAR + months;
+    BigInteger totalNanos = seconds.movePointRight(9).toBigInteger();
+    totalNanos =
+        totalNanos.add(BigInteger.valueOf(minutes * SECONDS_PER_MINUTE).multiply(NANOS_PER_SECOND));
+    totalNanos =
+        totalNanos.add(BigInteger.valueOf(hours * SECONDS_PER_HOUR).multiply(NANOS_PER_SECOND));
+
+    BigInteger totalMicros = totalNanos.divide(BigInteger.valueOf(NANOS_PER_MICRO));
+    BigInteger nanoFractions =
+        totalNanos.subtract(totalMicros.multiply(BigInteger.valueOf(NANOS_PER_MICRO)));
+
+    return Interval.builder()
+        .setMonths(Math.toIntExact(totalMonths))
+        .setDays(Math.toIntExact(days))
+        .setMicros(totalMicros.longValueExact())
+        .setNanoFractions(nanoFractions.shortValueExact())
+        .build();
+  }
+
+  /** Converts Interval to ISO8601 Duration Formatted String. */
+  public String ToISO8601() {
+    StringBuilder result = new StringBuilder();
+    result.append("P");
+
+    long months = this.months();
+    long years = months / MONTHS_PER_YEAR;
+    months = months - years * MONTHS_PER_YEAR;
+
+    if (years != 0) {
+      result.append(String.format("%dY", years));
+    }
+
+    if (months != 0) {
+      result.append(String.format("%dM", months));
+    }
+
+    if (this.days() != 0) {
+      result.append(String.format("%dD", this.days()));
+    }
+
+    BigInteger nanos = this.nanos();
+    BigInteger zero = BigInteger.valueOf(0);
+    if (nanos.compareTo(zero) != 0) {
+      result.append("T");
+      BigInteger hours = nanos.divide(NANOS_PER_HOUR);
+
+      if (hours.compareTo(zero) != 0) {
+        result.append(String.format("%sH", hours));
+      }
+
+      nanos = nanos.subtract(hours.multiply(NANOS_PER_HOUR));
+      BigInteger minutes = nanos.divide(NANOS_PER_MINUTE);
+      if (minutes.compareTo(zero) != 0) {
+        result.append(String.format("%sM", minutes));
+      }
+
+      nanos = nanos.subtract(minutes.multiply(NANOS_PER_MINUTE));
+      BigDecimal seconds = new BigDecimal(nanos).movePointLeft(9);
+
+      if (seconds.compareTo(new BigDecimal(zero)) != 0) {
+        result.append(String.format("%sS", seconds.stripTrailingZeros()));
+      }
+    }
+
+    if (result.length() == 1) {
+      result.append("0Y");
+    }
+
+    return result.toString();
+  }
+
+  /** Creates an interval which representing 0-duration. */
+  public static Interval zeroInterval() {
+    return builder().setMonths(0).setDays(0).setMicros(0).setNanoFractions((short) 0).build();
+  }
+
+  @Override
+  public boolean equals(Object rhs) {
+    if (!(rhs instanceof Interval)) {
+      return false;
+    }
+
+    Interval anotherInterval = (Interval) rhs;
+    return months() == anotherInterval.months()
+        && days() == anotherInterval.days()
+        && nanos().equals(anotherInterval.nanos());
```

Comment:
This appears to be missing a `micros` comparison. Which again also automatically leads me to believe that there must be some tests missing, as that would otherwise also have come up as test failures.

Reply:
nanos() here convert microseconds and nanoFractions to nanoseconds and comparing them.

I am doing this because I am allowing nanoFractions to be in range of [-999, 999].